### PR TITLE
add jsx-ignore option to no-magic-numbers rule

### PR DESCRIPTION
**Pull request by lizzzp1**
#### PR checklist

- [x] Addresses an existing issue: #4412
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

#### Overview of change:
Attempts to add an option to ignore jsx and leaves default as is.

#### Is there anything you'd like reviewers to focus on?
First PR in here, I may have missed some things. :)
Feedback appreciated.

#### CHANGELOG.md entry:

[new-rule-option] added `jsx-ignore` option to `no-magic-numbers` rule

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @lizzzp1! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: src/rules/noMagicNumbersRule.ts [hunk 1]
```diff
@@ -105,7 +115,8 @@ class NoMagicNumbersWalker extends Lint.AbstractWalker<number[]> {
         /* Using Object.is() to differentiate between pos/neg zero */
         if (
             !Rule.ALLOWED_NODES.has(node.parent.kind) &&
-            !this.options.some(allowedNum => Object.is(allowedNum, parseFloat(num)))
+            !this.options.some(allowedNum => Object.is(allowedNum, parseFloat(num))) &&
+            this.options.indexOf(IGNORE_JSX_OPTION) === -1
```
[the remark this answers is not in the document]
Comment (lizzzp1) [reply src/rules/noMagicNumbersRule.ts]: Ah. Got it, so it would ignore `ALL` jsx nodes. Missed that.

File: src/rules/noMagicNumbersRule.ts [hunk 2]
```diff
@@ -34,23 +45,36 @@ export class Rule extends Lint.Rules.AbstractRule {
             Magic numbers should be avoided as they often lack documentation.
             Forcing them to be stored in variables gives them implicit documentation.
         `,
-        optionsDescription: "A list of allowed numbers.",
+        optionsDescription: Lint.Utils.dedent`
+            \`${IGNORE_JSX_OPTION}\` specifies that jsx attributes with magic numbers should be ignored.
+            \`${ALLOWED_NUMBERS_OPTION}\` is a list of allowed numbers.`,
```
Comment (lizzzp1) [inline src/rules/noMagicNumbersRule.ts]: Might need some help on this wording 🙃

File: test/rules/no-magic-numbers/allowed-numbers/tslint.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "rules": {
+    "no-magic-numbers": [true, {
+      "allowed-numbers": [1337, 1337.7, -1337, -0]
```
Comment (JoshuaKGoldberg) [inline test/rules/no-magic-numbers/allowed-numbers/tslint.json]: Lol, I see what you did there...

File: src/rules/noMagicNumbersRule.ts [hunk 4]
```diff
@@ -101,11 +135,29 @@ class NoMagicNumbersWalker extends Lint.AbstractWalker<number[]> {
         return ts.forEachChild(sourceFile, cb);
     }
 
-    private checkNumericLiteral(node: ts.Node, num: string) {
+    private isAllowedNumber(num: string): boolean {
         /* Using Object.is() to differentiate between pos/neg zero */
+        const compareNumbers = (allowedNum: number) => Object.is(allowedNum, parseFloat(num));
+        return this.options.some(option => {
```
Comment (JoshuaKGoldberg) [inline src/rules/noMagicNumbersRule.ts]: > `this.options.some`

Now that the rule options are coming in as a more complex format, it's getting difficult to read through the rule and see how the options are meant to be used. Proposal: let's switch the responsibility of parsing options from the walker to the rule itself, such that the rule parses the options into an object of a known format & passes that object to the walker. Say, a `Set` of allowed numbers and a boolean for the new option?

`arrowReturnShorthandRule.ts` is a good simple case of this strategy, where the object passed to the walker just has the `multiline` property. `banRule.ts` has a complex object so it uses a `parseOptions` function.

File: src/rules/noMagicNumbersRule.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [reply src/rules/noMagicNumbersRule.ts]: Yeah I've always thought the rule's docs were a little terse... how about:

```ts
Options may either be a list of numbers to ignore (not consider 'magic'), or an object containing up to two properties:

* \`${ALLOWED_NUMBERS_OPTION}\` as the list of numbers to ignore
* \`${IGNORE_JSX_OPTION}\` to specify that 'magic' numbers should be allowed as JSX attributes
```

`alignRule.ts` iis an example of formatting & phrasing here, though it doesn't have the two formats...

Do you have the time to add `optionExamples`, such as how `arrowParensRule.ts` has them? Those are particularly helpful for understanding rule options. If so, your test cases look pretty good as copy & paste fodder for them.

File: src/rules/noMagicNumbersRule.ts (hunk 4 above)
Comment (lizzzp1) [reply src/rules/noMagicNumbersRule.ts]: Ok, that  makes sense!

File: src/rules/noMagicNumbersRule.ts [hunk 5]
```diff
@@ -104,13 +107,36 @@ export class Rule extends Lint.Rules.AbstractRule {
...
+                    ? parseOptions(this.ruleArguments)
+                    : parseOptions(Rule.DEFAULT_ALLOWED),
             ),
         );
     }
 }
 
-class NoMagicNumbersWalker extends Lint.AbstractWalker<Array<OptionsType | number>> {
+function parseOptions(options: Array<OptionsType | number>): ParsedOptionsType {
+    const parsedOptions: ParsedOptionsType = { [ALLOWED_NUMBERS_OPTION]: [] };
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/noMagicNumbersRule.ts]: Ha, funny. Nice catch.

File: src/rules/noMagicNumbersRule.ts [hunk 6]
```diff
@@ -70,13 +107,36 @@ export class Rule extends Lint.Rules.AbstractRule {
...
 }
 
-class NoMagicNumbersWalker extends Lint.AbstractWalker<number[]> {
+function parseOptions(options: Array<OptionsType | number>): ParsedOptionsType {
+    const parsedOptions: ParsedOptionsType = { [ALLOWED_NUMBERS_OPTION]: [] };
+    for (const option of options) {
+        if (typeof option === "number") {
+            parsedOptions[ALLOWED_NUMBERS_OPTION].push(option);
+        }
+        if (option.hasOwnProperty(ALLOWED_NUMBERS_OPTION)) {
```
[the remark this answers is not in the document]
Comment (lizzzp1) [reply src/rules/noMagicNumbersRule.ts]: oops! -- I'll update

File: src/rules/noMagicNumbersRule.ts [hunk 7]
```diff
@@ -120,16 +120,14 @@ function parseOptions(options: Array<OptionsType | number>): ParsedOptionsType {
...
+        } else if (option.hasOwnProperty(ALLOWED_NUMBERS_OPTION)) {
             const numberOptions = (option as AllowedNumbersType)[ALLOWED_NUMBERS_OPTION];
             if (Array.isArray(numberOptions) && numberOptions.length > 0) {
                 numberOptions.forEach((num: number) =>
                     parsedOptions[ALLOWED_NUMBERS_OPTION].push(num),
                 );
             }
-        }
-        if (option.hasOwnProperty(IGNORE_JSX_OPTION)) {
+        } else if (option.hasOwnProperty(IGNORE_JSX_OPTION)) {
```
Comment (lizzzp1) [inline src/rules/noMagicNumbersRule.ts]: Or is a switch statement of preference?

Comment (JoshuaKGoldberg) [reply src/rules/noMagicNumbersRule.ts]: How about a `continue` on the inside of the `if (typeof option === "number")` case?

It's possible an option object would have both `ALLOWED_NUMBERS_OPTION` and `IGNORE_JSX_OPTION`.

**Pull request merged**